polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1765,6 +1811,81 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
       LLVM_DEBUG(llvm::dbgs()
                  << "tryToAdapt: Could not infer captured accessor\n");
     }
+    if (isa<LLVM::LLVMArrayType>(expected) && collector.first != 0) {
+      auto arrayTy = cast<LLVM::LLVMArrayType>(expected);
+      size_t expectedNumElements = arrayTy.getNumElements();
+      Type expectedElemType = arrayTy.getElementType();
+      assert(expectedNumElements == collector.first && "Incomplete collector");
+      SmallVector<ArrayComponent> &offsetAndValues = collector.second;
+      // We need to sort them by index to ensure dominance.
+      llvm::sort(offsetAndValues,
+                 [](const ArrayComponent &p1, const ArrayComponent &p2) {
+                   return p1.offset < p2.offset;
+                 });
+      auto allConstant = llvm::all_of(offsetAndValues, [](const auto &p) {
+        return matchPattern(p.value, m_Constant());
+      });
+      // If all elements of the array are constant, we will insert a global
+      // array and capture the address of that global array containing the
+      // constant values of the array.
+      if (allConstant) {
+        builder.setInsertionPointToStart(
+            assigned.getDefiningOp()->getParentOfType<ModuleOp>().getBody());
```

Comment:
NIT: Inserting the operation right before the parent function might make testing easier